libraries/AP_DroneCAN/AP_DroneCAN.cpp

```diff
@@ -626,10 +634,17 @@ void AP_DroneCAN::SRV_send_esc(void)
 
         for (uint8_t i = esc_offset; i < max_esc_num && k < 20; i++) {
             if ((((uint32_t) 1) << i) & _esc_bm) {
-                // TODO: ESC negative scaling for reverse thrust and reverse rotation
-                float scaled = cmd_max * (hal.rcout->scale_esc_to_unity(_SRV_conf[i].pulse) + 1.0) / 2.0;
-
-                scaled = constrain_float(scaled, 0, cmd_max);
+                float scaled;
+                if ((((uint32_t) 1) << i) & _esc_rev_bm) {
+                    // min throttle is cmd_min (-8192), full speed reverse
+                    // mid throttle is 0, stopped
+                    scaled = cmd_max * _SRV_conf[i].output_norm;
```

Comment:
Andy is correct, this is because the individual servo ranges as set the the servo lib are not used for motors, instead we ignore them and use the `MOT_PWM_` values.

Reply:
Looking at the DShot implementation though I see that it's not relying on the ESC scaling and instead internally scales using an assumed 1000 to 2000 PWM value. I'd assumed that the ESC scaling was intended to allow the greater precision of DShot by calculating the source pulse data within a wider range. Is that not the case?

If it were the case, it would make sense potentially to support the greater precision in CAN too given we have ~16k steps available to us so converting from a range of 1000-2000 is losing quite a lot of precision. For a rover (and probably most things) I can well imagine this is moot, but architecturally I suppose the system shouldn't forever be tied to PWM precision.

Reply:
I'm not sure this is a problem people have right now, so although I agree in theory in practice I suspect the churn is not worth it

Reply:
Yeah indeed. And while newer ESC protocols have this increased precision without much cost due to high speed buses, the RC protocols still tend to have the lower precision anyway.

I find it interesting having entered Ardupilot 3 years ago in a mostly post-PWM world (at least on the RC and ESC side) and finding everything pandering to patterns that had essentially already been superseded was interesting. There may come a time a decade down the road where PWM is not the default nomenclature...

Though my planes are mostly PWM I suppose :grin: 

Thanks for the review.